File: src/main/java/io/github/wntopia/gikipedia/server/global/config/DataGsmOAuthEnvironment.java
package io.github.wntopia.gikipedia.server.global.config;

import java.util.Map;
import org.springframework.boot.context.properties.ConfigurationProperties;

@ConfigurationProperties(prefix = "spring.security.oauth2.client")
public record DataGsmOAuthEnvironment(
    Map<String, Registration> registration, Map<String, Provider> provider) {
  private static final String DATAGSM = "datagsm";
  private static final String AUTHORIZATION_PATH = "/v1/oauth/authorize";
  private static final String USER_INFO_PATH = "/userinfo";

  public String authorizationBaseUrl() {
    return removeSuffix(datagsmProvider().authorizationUri(), AUTHORIZATION_PATH);
  }

  public String resourceBaseUrl() {
    return removeSuffix(datagsmProvider().userInfoUri(), USER_INFO_PATH);
  }

  public String clientId() {
    return datagsmRegistration().clientId();
  }

  public String clientSecret() {
    return datagsmRegistration().clientSecret();
  }

  public String redirectUri() {
    return datagsmRegistration().redirectUri();
  }

  public String scope() {
    return datagsmRegistration().scope();
  }

  public String successRedirectUri() {
    return datagsmRegistration().successRedirectUri();
  }

  public String failureRedirectUri() {
    return datagsmRegistration().failureRedirectUri();
  }

  private Registration datagsmRegistration() {
    if (registration == null || registration.get(DATAGSM) == null) {
      throw new IllegalStateException(
          "spring.security.oauth2.client.registration.datagsm must be configured");
    }
    return registration.get(DATAGSM);
  }

  private Provider datagsmProvider() {
    if (provider == null || provider.get(DATAGSM) == null) {
      throw new IllegalStateException(
          "spring.security.oauth2.client.provider.datagsm must be configured");
    }
    return provider.get(DATAGSM);
  }

  private static String removeSuffix(String value, String suffix) {
    if (value == null || !value.endsWith(suffix)) {
      return value;
    }
    return value.substring(0, value.length() - suffix.length());
  }

  public record Registration(
      String clientId,
      String clientSecret,
      String redirectUri,
      String scope,
      String successRedirectUri,
      String failureRedirectUri) {}

  public record Provider(String authorizationUri, String tokenUri, String userInfoUri) {}
}

File: src/main/kotlin/io/github/wntopia/gikipedia/server/global/config/DataGsmOAuthEnvironment.kt
package io.github.wntopia.gikipedia.server.global.config

import org.springframework.boot.context.properties.ConfigurationProperties

@ConfigurationProperties(prefix = "spring.security.oauth2.client")
data class DataGsmOAuthEnvironment(
    val registration: Map<String, Registration>?,
    val provider: Map<String, Provider>?,
) {
    fun authorizationBaseUrl(): String? = removeSuffix(datagsmProvider().authorizationUri, AUTHORIZATION_PATH)

    fun resourceBaseUrl(): String? = removeSuffix(datagsmProvider().userInfoUri, USER_INFO_PATH)

    fun clientId(): String? = datagsmRegistration().clientId

    fun clientSecret(): String? = datagsmRegistration().clientSecret

    fun redirectUri(): String? = datagsmRegistration().redirectUri

    fun scope(): String? = datagsmRegistration().scope

    fun successRedirectUri(): String? = datagsmRegistration().successRedirectUri

    fun failureRedirectUri(): String? = datagsmRegistration().failureRedirectUri

    private fun datagsmRegistration(): Registration =
        registration?.get(DATAGSM)
            ?: throw IllegalStateException(
                "spring.security.oauth2.client.registration.datagsm must be configured",
            )

    private fun datagsmProvider(): Provider =
        provider?.get(DATAGSM)
            ?: throw IllegalStateException(
                "spring.security.oauth2.client.provider.datagsm must be configured",
            )

    private fun removeSuffix(
        value: String?,
        suffix: String,
    ): String? =
        if (value == null || !value.endsWith(suffix)) {
            value
        } else {
            value.substring(0, value.length - suffix.length)
        }

    data class Registration(
        val clientId: String?,
        val clientSecret: String?,
        val redirectUri: String?,
        val scope: String?,
        val successRedirectUri: String?,
        val failureRedirectUri: String?,
    )

    data class Provider(
        val authorizationUri: String?,
        val tokenUri: String?,
        val userInfoUri: String?,
    )

    companion object {
        private const val DATAGSM = "datagsm"
        private const val AUTHORIZATION_PATH = "/v1/oauth/authorize"
        private const val USER_INFO_PATH = "/userinfo"
    }
}
